images/virtualization-artifact/pkg/controller/vmip/vmip_webhook.go

```diff
@@ -0,0 +1,156 @@
+	if !ok {
+		return nil, fmt.Errorf("expected a new VirtualMachineIPAddress but got a %T", obj)
+	}
+
+	v.log.Info("Validate VirtualMachineIP creating", "name", vmip.Name, "type", vmip.Spec.Type, "address", vmip.Spec.StaticIP)
+
+	err := v.validateSpecFields(vmip.Spec)
+	if err != nil {
+		return nil, fmt.Errorf("error validating VirtualMachineIP creation: %w", err)
+	}
+
+	if vmip.Spec.Type == v1alpha2.VirtualMachineIPAddressTypeStatic {
+		ip := vmip.Spec.StaticIP
+
+		if ip != "" {
+			allocatedIPs, err := util.GetAllocatedIPs(ctx, v.client, vmip.Spec.Type)
+			if err != nil {
+				return nil, fmt.Errorf("error getting allocated ips: %w", err)
+			}
+
+			allocatedLease, ok := allocatedIPs[ip]
+			if ok && allocatedLease.Spec.VirtualMachineIPAddressRef != nil &&
+				(allocatedLease.Spec.VirtualMachineIPAddressRef.Namespace != vmip.Namespace ||
+					allocatedLease.Spec.VirtualMachineIPAddressRef.Name != vmip.Name) {
+				return nil, fmt.Errorf("VirtualMachineIPAddress cannot be created: the IP address %s has already been allocated by VirtualMachineIPAddress/%s in ns/%s", ip, allocatedLease.Spec.VirtualMachineIPAddressRef.Name, allocatedLease.Spec.VirtualMachineIPAddressRef.Namespace)
+			}
+		}
+	}
+
+	if vmip.Spec.Type == v1alpha2.VirtualMachineIPAddressTypeAuto && vmip.Spec.StaticIP != "" {
+		v.log.Error(nil, "Invalid combination: StaticIP is specified with Type 'Auto'. The StaticIP will be ignored.", "VirtualMachineIPName", vmip.Name, "StaticIP", vmip.Spec.StaticIP)
+	}
+
+	return nil, nil
+}
+
+func (v *Validator) ValidateUpdate(_ context.Context, oldObj, newObj runtime.Object) (admission.Warnings, error) {
+	oldVmip, ok := oldObj.(*v1alpha2.VirtualMachineIPAddress)
+	if !ok {
+		return nil, fmt.Errorf("expected an old VirtualMachineIP but got a %T", oldObj)
+	}
+
+	newVmip, ok := newObj.(*v1alpha2.VirtualMachineIPAddress)
+	if !ok {
+		return nil, fmt.Errorf("expected a new VirtualMachineIP but got a %T", newObj)
+	}
+
+	v.log.Info("Validate VirtualMachineIP updating", "name", newVmip.Name,
+		"old.type", oldVmip.Spec.Type, "new.type", newVmip.Spec.Type,
+		"old.address", oldVmip.Spec.StaticIP, "new.address", newVmip.Spec.StaticIP,
+	)
+
+	boundCondition, exist := service.GetCondition(vmipcondition.BoundType, oldVmip.Status.Conditions)
+	if exist && boundCondition.Status == metav1.ConditionTrue {
+		if oldVmip.Spec.Type == v1alpha2.VirtualMachineIPAddressTypeAuto && newVmip.Spec.Type == v1alpha2.VirtualMachineIPAddressTypeStatic {
+			v.log.Info("Change the VirtualMachineIP address type to 'Auto' from 'Static' for ip: ", "address", newVmip.Spec.StaticIP)
+			if newVmip.Spec.StaticIP != newVmip.Status.Address {
+				return nil, fmt.Errorf("only type change Auto->Static is allowed: can't change current IP %s to the specified %s", newVmip.Status.Address, newVmip.Spec.StaticIP)
+			}
+
+			err := v.validateSpecFields(newVmip.Spec)
+			if err != nil {
+				return nil, fmt.Errorf("error validating VirtualMachineIP updating: %w", err)
```

Comment:
"error validating VirtualMachineIP update"

Reply:
I corrected it in several places.